Run `Standard.Google` in dual JVM mode

JaroslavTulach opened the pull request:
### Pull Request Description

- Turning on #13172 for `Standard.Google`
- [Allows accessing Enso language](https://github.com/enso-org/enso/pull/14040/files#r2415572399) from _dual JVM mode_
- Introduces [Problem.problemType()](https://github.com/enso-org/enso/pull/14040/files#r2415575046) string representation
- Support for [debugging of Enso build system](https://github.com/enso-org/enso/pull/14040/files#r2412828748) written in Rust


### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
- [x] Google [tested are executed without errors](https://github.com/enso-org/enso/pull/14040#issuecomment-3384086781)
- [x] check yet another [nightly release](https://github.com/enso-org/enso/actions/runs/19256839216)
    - Linux?
- [x] Standard libraries [benchmark run](https://github.com/enso-org/enso/actions/runs/18365719675) is OK, another [run](https://github.com/enso-org/enso/actions/runs/19305839350)
    - [one more run](https://github.com/enso-org/enso/actions/runs/19328182649) with #14238 fix in

JaroslavTulach committed:
Run Standard.Google in dual JVM mode

JaroslavTulach commented:
- @AdRiley, I am trying to verify this change is correct. However doing:
```bash
corepack pnpm i
corepack pnpm compile
corepack pnpm dev:gui
```
- connecting with a browser and then trying https://github.com/enso-org/enso/tree/16125eb43250e9eadfc4dc75b74e639d14d89d71/test/Google_Test doesn't work.

- I am now trying to generate AppImage via `./run ide build` script. With such an IDE I am able to create the credentials:

<img width="655" height="408" alt="image" src="https://github.com/user-attachments/assets/414e7543-5f5b-41ca-b920-05e95cb15990" />

- and then I can run (and fail) the tests:
```
enso$ ENSO_LAUNCHER=native sbt buildEngineDistribution

    - [FAILED] can read multiple sheets - error case - With_New_Column - no valid sheets [297ms]
        Reason: An unexpected panic was thrown: (Unsupported_Argument_Types.Error [] 'Cannot convert \'CredentialReference[secretId=secret-3360Yd1PR0XaCGaifwwFOJrqb3d, serviceName
    - [FAILED] can read multiple sheets - error case - With_New_Column - no valid sheets [297ms]
        Reason: An unexpected panic was thrown: (Unsupported_Argument_Types.Error [] 'Cannot convert \'CredentialReference[secretId=secret-3360Yd1PR0XaCGaifwwFOJrqb3d, serviceName=Google]\'(language: Java, type: com.oracle.truffle.polyglot.PolyglotMap) to Java type \'org.enso.base.enso_cloud.ExternalLibraryCredentialHelper$CredentialReference\': Unsupported target type.')
        at <enso> case_branch(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets_Workbook.enso:177:13-93)
        at <enso> Google_Sheets_Workbook._wrap_credentials(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets_Workbook.enso:175-182)
        at <enso> Google_Sheets_Workbook.type.new<arg-1>(/home/devel/NetBeansPro
    - [FAILED] can read multiple sheets - error case - With_New_Column - no valid sheets [297ms]
        Reason: An unexpected panic was thrown: (Unsupported_Argument_Types.Error [] 'Cannot convert \'CredentialReference[secretId=secret-3360Yd1PR0XaCGaifwwFOJrqb3d, serviceName=Google]\'(language: Java, type: com.oracle.truffle.polyglot.PolyglotMap) to Java type \'org.enso.base.enso_cloud.ExternalLibraryCredentialHelper$CredentialReference\': Unsupported target type.')
        at <enso> case_branch(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets_Workbook.enso:177:13-93)
        at <enso> Google_Sheets_Workbook._wrap_credentials(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets_Workbook.enso:175-182)
        at <enso> Google_Sheets_Workbook.type.new<arg-1>(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets_Workbook.enso:45:47-75)
        at <enso> Google_Sheets_Workbook.type.new(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets_Workbook.enso:45:19-76)
        at <enso> Google_Sheets.read(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets.enso:15:9-57)
        at <enso> Main.add_specs.Main.add_specs<arg-1>(/home/devel/NetBeansProjects/enso/enso/test/Google_Test/src/Main.enso:311:18-115)
        at <enso> case_branch.case <internal-1178>(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Test/0.0.0-dev/src/Group.enso:37:61-64)
        at <enso> Helpers.execute_spec_code<arg-1>(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Test/0.0.0-dev/src/Helpers.enso:56:18-34)
   jects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets_Workbook.enso:45:47-75)
        at <enso> Google_Sheets_Workbook.type.new(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets_Workbook.enso:45:19-76)
        at <enso> Google_Sheets.read(/home/devel/NetBeansProjects/enso/enso/distrib
    - [FAILED] can read multiple sheets - error case - With_New_Column - no valid sheets [297ms]
        Reason: An unexpected panic was thrown: (Unsupported_Argument_Types.Error [] 'Cannot convert \'CredentialReference[secretId=secret-3360Yd1PR0XaCGaifwwFOJrqb3d, serviceName=Google]\'(language: Java, type: com.oracle.truffle.polyglot.PolyglotMap) to Java type \'org.enso.base.enso_cloud.ExternalLibraryCredentialHelper$CredentialReference\': Unsupported target type.')
        at <enso> case_branch(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets_Workbook.enso:177:13-93)
        at <enso> Google_Sheets_Workbook._wrap_credentials(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets_Workbook.enso:175-182)
        at <enso> Google_Sheets_Workbook.type.new<arg-1>(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets_Workbook.enso:45:47-75)
        at <enso> Google_Sheets_Workbook.type.new(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets_Workbook.enso:45:19-76)
        at <enso> Google_Sheets.read(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets.enso:15:9-57)
        at <enso> Main.add_specs.Main.add_specs<arg-1>(/home/devel/NetBeansProjects/enso/enso/test/Google_Test/src/Main.enso:311:18-115)
        at <enso> case_branch.case <internal-1178>(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Test/0.0.0-dev/src/Group.enso:37:61-64)
        at <enso> Helpers.execute_spec_code<arg-1>(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Test/0.0.0-dev/src/Helpers.enso:56:18-34)
   ution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets.enso:15:9-57)
        at <enso> Main.add_specs.Main.add_specs<arg-1>(/home/devel/NetBeansProjects/enso/enso/test/Google_Test/src/Main.enso:311:18-115)
        at <enso> case_branch.case <internal-1178>(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Test/0.0.0-dev/src/Group.enso:37:61-64)
        at <enso> Helpers.execute_spec_code<arg-1>(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Test/0.0.0-dev/src/Helpers.enso:56:18-34)
   =Google]\'(language: Java, type: com.oracle.truffle.polyglot.PolyglotMap) to Java type \'org.enso.base.enso_cloud.ExternalLibraryCredentialHelper$CredentialReference\': Unsupported target type.')
        at <enso> case_branch(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets_Workbook.enso:177:13-93)
        at <enso> Google_Sheets_Workbook._wrap_credentials(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets_Workbook.enso:175-182)
        at <enso> Google_Sheets_Workbook.type.new<arg-1>(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets_Workbook.enso:45:47-75)
        at <enso> Google_Sheets_Workbook.type.new(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets_Workbook.enso:45:19-76)
        at <enso> Google_Sheets.read(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets.enso:15:9-57)
        at <enso> Main.add_specs.Main.add_specs<arg-1>(/home/devel/NetBeansProjects/enso/enso/test/Google_Test/src/Main.enso:311:18-115)
        at <enso> case_branch.case <internal-1178>(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Test/0.0.0-dev/src/Group.enso:37:61-64)
        at <enso> Helpers.execute_spec_code<arg-1>(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Test/0.0.0-dev/src/Helpers.enso:56:18-34)
```
- that seems to be a _dual JVM like problem_. 
- similar to #14039 
- updating to 3c0c853 (with fix for #14039) doesn't help, CCing @jdunkerley

JaroslavTulach committed:
Merge remote-tracking branch 'origin/develop' into wip/jtulach/DualGoogle

JaroslavTulach committed:
Merge remote-tracking branch 'origin/develop' into wip/jtulach/DualGoogle

JaroslavTulach commented:
I have all the setup to execute the tests in regular JVM mode:
```
enso$ sbt
sbt:enso> runEngineDistribution --run test/Google_Test
48 tests succeeded.
0 tests failed.
0 tests skipped.
0 groups skipped.
[success] Total time: 102 s (01:42), completed 7. 10. 2025 11:12:50
```
now onto running the same tests in _dual JVM mode_. With bc03db6 there is eleven remaining failures:
```
sbt:enso> runEngineDistribution --run test/Google_Test 
  --vm.D=polyglot.enso.classLoading=Standard.Google:guest,enso_dev.Google_Test:guest,hosted

37 tests succeeded.
11 tests failed.
0 tests skipped.
0 groups skipped.
[success] Total time: 105 s (01:45), completed 7. 10. 2025 17:18:47
```
we are on track to get the _dual JVM mode_ working for `Standard.Google` library!

JaroslavTulach committed:
Pass the right CredentialReference into as_credential_reference via an argument

JaroslavTulach committed:
Provide own ProblemAggregator

JaroslavTulach committed:
Updating documentation of running the test/Google_Test

JaroslavTulach committed:
Obtain Enso bindings in dual JVM mode by sending an OtherJvmMessage

JaroslavTulach committed:
Using OtherJvm support is no error!

JaroslavTulach committed:
Recognize Problem types by simple class name so values from different JVMs have the same ID

JaroslavTulach committed:
Rather than wrapping Java exceptions throw Enso exceptions directly from Java code

JaroslavTulach commented on std-bits/google/src/main/java/org/enso/google/GoogleSheetsForEnso.java:
```diff
@@ -63,7 +63,11 @@ public Table getSheetRange(
+      var emptySheetError = emptySheetType.invokeMember("Error");
+      var errorType = EnsoMeta.getType("Standard.Base.Error", "Error");
+      var error = errorType.invokeMember("throw", emptySheetError);
+      throw error.throwException();
```
- cdbf8fb35f8e446fdff03594c9b68c2c7cd6f59e is a solution to [dual exceptions problem](https://github.com/enso-org/enso/pull/14037#discussion_r2396959869), @AdRiley 
- rather than throwing Java exception and then trying to convert it to `Empty_Sheet.Error`
- let's **directly throw** `Empty_Sheet.Error`!
- the `getSheetRange` method returns `Table` - as such we need to throw it as `Panic` (hence `error.throwException()`
   - otherwise just `return error;` to propagate the value as (dataflow) `Error`
- because we throw `Panic` we have `Panic.recover` in the Enso code
- if we returned (dataflow) `Error` - we could _delete the Enso conversion code_ completely!

JaroslavTulach replied:
- we can go even further with this approach, @jdunkerley
- if we just `return error;` then there is no need to do anything special on the Enso side
- as de343f3 illustrates, it is possible to remove `Empty_Sheet.handle_java_exception`
- the only drawback of the de343f3 is the [change of function return type](https://github.com/enso-org/enso/pull/14040#discussion_r2413035000) to `Object`
   - we need to return either `Table` or `Value` and `Object` is the only common superclass
- the `throw error.throwException()` solution workarounds this by _returning via an exception_ (at the cost of necessary handler on the Enso side)

JaroslavTulach committed:
EmptySheetException has been removed

JaroslavTulach commented:
Right now we have:
- failing [JUnit tests](https://github.com/enso-org/enso/actions/runs/18324396747/job/52186049270?pr=14040#step:14:6691) because of changes to `DataflowError.throwException()`:
```
[error] (runtime-language-epb / Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] (runtime-integration-tests / Test / test) sbt.TestsFailedException: Tests unsuccessful
```
- one [API change](https://github.com/enso-org/enso/actions/runs/18324396747/job/52186049279?pr=14040#step:14:1027) because of adding default Java class argument to `as_credential_reference`
- intermittent(?) [stalled(?) test](https://github.com/enso-org/enso/actions/runs/18324396747/job/52188264599?pr=14040#step:14:5281) on Windows

Other than that we are ready to move on. Heuréka!

JaroslavTulach committed:
Print out exact sbt command to regenerate caches

JaroslavTulach commented on build_tools/build/src/engine/context.rs:
```diff
@@ -645,23 +646,27 @@ impl RunContext {
                 error!("Current API vs Old API: {}", err);
-                error!("If you wish to overwrite the current API in the directory {}, run the following command {},
+                error!("If you wish to overwrite the current API in the directory {}, run the following command 
+                       sbt \"runEngineDistribution --no-ir-caches --docs=api --in-project {}\"
```
This change, @AdRiley, @GregoryTravis, @jdunkerley, will print:
```
index 8b63b3d24d..f2c90e184d 100644
--- home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Table/0.0.0-dev/docs/api/Errors.md
+++ home/devel/NetBeansProjects/enso/enso/built-distribution/enso-engine-2025.3.1-dev-linux-amd64/enso-2025.3.1-dev/lib/Standard/Table/2025.3.1-dev/docs/api/Errors.md
@@ -43,7 +43,7 @@
     - to_display_text self -> Standard.Base.Any.Any
 - type Empty_Sheet
     - Error
-    - handle_java_exception -> Standard.Base.Any.Any
+    - handle_java_exception ~action:Standard.Base.Any.Any -> Standard.Base.Any.Any
     - to_display_text self -> Standard.Base.Any.Any
 - type Existing_Column
     - Error column_name:Standard.Base.Any.Any

ERROR main_internal: enso_build::engine::context: If you wish to overwrite the current API in the directory distribution/lib/Standard/Table/0.0.0-dev/docs/api, run the following command 
  sbt "runEngineDistribution --no-ir-caches --docs=api --in-project distribution/lib/Standard/Table/0.0.0-dev"
  and commit the modified files
ERROR main_internal: enso_build_cli: error=API check failed for library Standard.Table
 INFO main_internal: enso_build_cli: close
Error: API check failed for library Standard.Table
```
in case of API check failure. Which I can then directly copy and run the regenerator:
```bash
enso$ sbt "runEngineDistribution --no-ir-caches --docs=api --in-project distribution/lib/Standard/Table/0.0.0-dev"
```
Please note the `--no-ir-caches` flag added for your convenience as _the actionable_ applied after recent complains.

JaroslavTulach committed:
Provide credential_reference Java class as defaulted argument

JaroslavTulach committed:
Let's rely on currying instead of passing the argument thru

JaroslavTulach commented on .vscode/launch.json:
```diff
@@ -41,6 +41,16 @@
+      "name": "Rust ./run",
+      "program": "${workspaceFolder}/target/rust/buildscript/enso-build-cli",
+      // provide arguments, for example: 
+      // "args": ["backend", "stdlib-api-check"]
```
With c4174e5d8 it should be easy to debug `./run` executable in [VSCode with Rust Analyzer](https://github.com/enso-org/enso/pull/14037/files#r2393592162):

<img width="1211" height="364" alt="debug ./run" src="https://github.com/user-attachments/assets/78906bd4-7436-410f-b94c-f1f2f66cb748" />

Change the `args` list and press F5, @Akirathan, @4e6. This is a continuation of the effort started in https://github.com/enso-org/enso/pull/14037/files#r2393592162

JaroslavTulach committed:
Use Path to avoid cargo linting

JaroslavTulach committed:
No need for Empty_Sheet.handle_java_exception when returning Error.throw result

JaroslavTulach commented on std-bits/google/src/main/java/org/enso/google/GoogleSheetsForEnso.java:
```diff
@@ -41,7 +41,7 @@ public static GoogleSheetsForEnso create(WrappedGoogleCredentials credentials)
   }
 
-  public Table getSheetRange(
+  public Object getSheetRange(
```
- need a common super type of `Table` and `Value` because of [returning Error.throw](https://github.com/enso-org/enso/pull/14040#discussion_r2413027910)

jdunkerley replied:
Rather than this change - lets add a `org.enso.google.EmptySheetException` and throw that instead and catch as we were before. This will be in the correct JVM.

We can then separate the changes for Google sheet from changes to the Excel library.

We can then look to use a more general DataFlow Error throwing as a solution.

jdunkerley replied:
Alternatively if we inline the Panic.catch then it will catch the correct exception.

JaroslavTulach replied:
This change should be reverted as of ec9abedc5f5bea1369812c5a6e6659d7e5e1327f

JaroslavTulach committed:
Propagate stacktraces when re-throwing DataflowError as PanicException

JaroslavTulach committed:
Check behavior of Error.throw invoked from polyglot Java

JaroslavTulach committed:
Convert proxy class to real StorageType

JaroslavTulach committed:
Delay call to context.getEnv().getScopePublic until access to Enso language is needed

JaroslavTulach commented:
> ```
> sbt:enso> runEngineDistribution --run test/Google_Test 
>   --vm.D=polyglot.enso.classLoading=Standard.Google:guest,enso_dev.Google_Test:guest,hosted
> 

There is now a last remaining failure:
```
    - [FAILED] can read multiple sheets - one sheet [2192ms]
        Reason: An unexpected panic was thrown: (Unsupported_Argument_Types.Error [OtherJvmObject{id=1537}, true, , JavaObject[org.enso.table.problems.ProblemAggregator@6fa60e24 (org.enso.table.problems.ProblemAggregator)]] 'Cannot convert \'org.enso.table.data.table.Table@684bfebc\'(language: Enso Polyglot Bridge, type: org.enso.table.data.table.Table) to Java type \'org.enso.table.data.table.Table\': Unsupported target type.')
        at <enso> In_Memory_Table_Implementation.zip.new_java_table(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Table/0.0.0-dev/src/Internal/In_Memory_Table_Implementation.enso:555:17-143)
        at <enso> Java_Problems.with_problem_aggregator(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Java_Problems.enso:120:14-25)
        at <enso> In_Memory_Table_Implementation.zip<arg-1>(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Table/0.0.0-dev/src/Internal/In_Memory_Table_Implementation.enso:554-555)
        at <enso> case_branch(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Base/0.0.0-dev/src/Errors/Problem_Behavior.enso:88-89)
        at <enso> Problem_Behavior.attach_problems_before(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Base/0.0.0-dev/src/Errors/Problem_Behavior.enso:84-92)
        at <enso> In_Memory_Table_Implementation.zip(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Table/0.0.0-dev/src/Internal/In_Memory_Table_Implementation.enso:553-556)
        at <enso> Table.zip(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Table/0.0.0-dev/src/Table.enso:3113:9-93)
        at <enso> case_branch<arg-2>(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets_Workbook.enso:114:21-124)
        at <enso> case_branch(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets_Workbook.enso:110-114)
        at <enso> Google_Sheets_Workbook.read_many(/home/devel/NetBeansProjects/enso/enso/distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets_Workbook.enso:102-114)
        at <enso> Main.add_specs.Main.add_specs<arg-1>(/home/devel/NetBeansProjects/enso/enso/test/Google_Test/src/Main.enso:223:18-48)
```
it should probably be mitigated by using the same technique as in https://github.com/enso-org/enso/pull/14037/commits/51366e2f4072388cbed527dd9f8a66bbbbcc1869

JaroslavTulach committed:
Convert internal_java_table to desired Java class when needed

JaroslavTulach commented:
> ... mitigated by using the same technique - **done** in 2a39d26bb1

```
sbt:enso> runEngineDistribution --run test/Google_Test --vm.D=polyglot.enso.classLoading=Standard.Google:guest,enso_dev.Google_Test:guest,hosted

48 tests succeeded.
0 tests failed.
0 tests skipped.
0 groups skipped.

[success] Total time: 111 s (01:51), completed 9. 10. 2025 6:18:10
```
I believe we are ready! @AdRiley please review and verify my claims. Thank you.

JaroslavTulach commented on distribution/lib/Standard/Table/0.0.0-dev/src/In_Memory_Table.enso:
```diff
@@ -35,9 +36,14 @@ type In_Memory_Table
+        raw_table = self.internal_java_table
+        if raw_table.is_a table_class then raw_table else
+            # coming from the other JVM and deserves conversion
+            new_columns = raw_table.getColumns.map c-> column_class.new c.getName c.getStorage
```
- using same technique as in https://github.com/enso-org/enso/pull/14037/commits/51366e2f4072388cbed527dd9f8a66bbbbcc1869
- e.g. converting via `ColumnStorage` (which is an interface)
- and creating new `Table` for `Column` (of the right class) delegating to proxy of the interface

JaroslavTulach commented on std-bits/base/src/main/java/org/enso/base/polyglot/EnsoMeta.java:
```diff
@@ -6,7 +6,13 @@
   private static Value getBindings() {
-    return Context.getCurrent().getBindings("enso");
+    var ctx = Context.getCurrent();
+    var bindings = ctx.getPolyglotBindings().getMember("ensoBindings");
```
- With these changes we can now access Enso classes even from _dual JVM_

JaroslavTulach commented on distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Java_Problems.enso:
```diff
@@ -37,31 +41,31 @@ polyglot java import org.enso.table.util.problems.InvalidNames
    Convert a Java problem into its Enso equivalent.
-translate_problem p = case p of
-    _ : InvalidAggregation ->
+translate_problem p = case p.problemType of
```
Introduction of `Problem.problemType()` allow us to `translate_problem` regardless from which JVM it origins

jdunkerley replied:
Not sure I like this approach, and wonder if each Problem should be responsible for making an Enso object.

JaroslavTulach replied:
- this is 1:1 rewrite
   - it is easy to check it does the same as original code
   - making _"Problem should be responsible for making an Enso object"_ is a bigger and non trivial change
- I don't like this approach either
   - in my opinion the Java code should use `EnsoMeta` to directly instantiate the right Enso classes
   - then there would be no need for these conversions
- e.g. using `EnsoMeta` as #14287 demonstrates

JaroslavTulach commented on test/Google_Test/README.md:
````diff
@@ -2,10 +2,30 @@ This is a set of tests for the Google integration for Enso.
+
+```bash
+enso$ ./run ide build
+enso$ ls ./dist/ide/enso-linux-*.AppImage
````
- this would be very nice to automate, @otulach, @PabloBuchu!
- first of all we need to run the IDE and create the `GoogleSheet` credential via playwright
- then we need to execute the `Google_Test`s from CLI

AdRiley commented:
<img width="1248" height="530" alt="image" src="https://github.com/user-attachments/assets/03565e0d-3a45-47ec-826d-ed7632363937" />

Testing Google Sheets is showing an error...

JaroslavTulach commented:
> Testing Google Sheets is showing an error...

- Yes, that's typical dual JVM problem.
- I used a fix like this at one place: https://github.com/enso-org/enso/commit/2a39d26bb129569060a0f676056f281fdd25e4c9
- The fix is of this problem is likely to be the same. Just somewhere else.

AdRiley commented:
Further testing indicates this error only happens in the IDE. If we extend the test to

<img width="854" height="558" alt="image" src="https://github.com/user-attachments/assets/844df1cf-e080-4342-a53a-9893ea1f606b" />

and then run from command line. We get back the data from google sheets as we expect

<img width="1475" height="332" alt="image" src="https://github.com/user-attachments/assets/7f62a34a-f80e-4751-bac6-9bb952ccee96" />

AdRiley commented:
Updating the test project to this

    google_sheets_workbook1 = Google_Sheets.read '1yKs-vGmxpo5oEKzhk7HN2xF40ux1q07RuzZn_7PQDDI' (Enso_Secret.get 'enso://Users/AdRiley/GS')
    node1 = google_sheets_workbook1.read 'Sheet2' ..No_Headers
    # node1 = Table.new [['A', ['S2A']], ['B', ['S2B']], ['C', ['S2C']]]
    node4 = node1.pretty
    node3 = IO.println node4
    test = Standard.Visualization.Table.Visualization.prepare_visualization node1
    IO.println test
    
    Causes a new error
        
<img width="1692" height="547" alt="image" src="https://github.com/user-attachments/assets/a9654772-837c-44eb-81db-0df1a602a42c" />

JaroslavTulach committed:
Merging with develop with GraalVM 25

JaroslavTulach commented:
> Updating the test project to this

Thanks @AdRiley for the testing. Now, we other tasks like #14019 being integrated, let's move this PR on. I am getting following error:

<img width="1139" height="82" alt="obrazek" src="https://github.com/user-attachments/assets/898ec5ab-c4fa-490c-bda9-4d2f2c4f1d5e" />

when trying your code. Obviously, I don't have access to your credentials. But on a general note: How can projects referencing secrets be made transferable between different people?

JaroslavTulach committed:
Properly report when createPolyglotJava fails

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/EnsoPolyglotJava.java:
```diff
@@ -153,10 +153,19 @@ static void close(EnsoContext ctx) {
+    try {
+      polyglotJava = createPolyglotJava(ctx);
+    } catch (Throwable t) {
+      logger.log(Level.ERROR, "Cannot create polyglotJava", t);
```
- After merging the most recent `develop` in
- and especially #13178 
- we are now getting
```
INFO: Initializing OtherJvm support!
SEVERE: Cannot create polyglotJava
java.lang.AssertionError: Error creating JVM: -5
       at org.enso.jvm.channel/org.enso.jvm.channel.JVM.initializeEnv(JVM.java:114)
       at org.enso.jvm.channel/org.enso.jvm.channel.JVM.env(JVM.java:87)
       at org.enso.jvm.channel/org.enso.jvm.channel.Channel.create(Channel.java:161)
       at org.enso.jvm.interop/org.enso.jvm.interop.api.OtherJvmClassLoader.createImpl(OtherJvmClassLoader.java:129)
       at org.enso.jvm.interop/org.enso.jvm.interop.api.OtherJvmClassLoader.create(OtherJvmClassLoader.java:76)
       at org.enso.interpreter.epb.JavaPolyglotNode.create(JavaPolyglotNode.java:21)
       at org.enso.interpreter.epb.ForeignEvalNode.parseJava(ForeignEvalNode.java:124)
```
- when trying to initialize the HotSpot JVM here
- because there already is one HotSpot JVM in the process (created by #13178)
- and the [documentations says](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/invocation.html) and continues to [say it in the JDK 25](https://docs.oracle.com/en/java/javase/25/docs/specs/jni/invocation.html#jni_getcreatedjavavms) as well:

> Creation of multiple VMs in a single process is not supported. 

- one should probably read the specification more carefully before getting too optimistic

JaroslavTulach replied:
> > Creation of multiple VMs in a single process is not supported.
> 
>     * one should probably read the specification more carefully before getting too optimistic

- with #14232 this problem will be solved
- by loading `ydoc-server` by a _native JVM_ leaving HotSpot JVM for us

JaroslavTulach committed:
Merging with ydoc-server in native image changes

JaroslavTulach commented:
### Notes to myself about my recent findings

- I've updated this PR to latest `develop` in 0534128
- e.g. to a state when it merges with #14233
- "nothing works" - I cannot really open @AdRiley [project](https://github.com/enso-org/enso/pull/14040#issuecomment-3401845422)
- hence in 638ba98 I am "downgrading" to include just #14232

### Does it work on 25.0.0?

```
$ PATH=$HOME/bin/graalvm-community-openjdk-25+25.1/bin:$PATH
$ sbt buildEngineDistribution
$ corepack pnpm run dev:gui
```
e.g. _full JVM mode_ loads the project OK: 
<img width="1172" height="647" alt="obrazek" src="https://github.com/user-attachments/assets/3a3ee521-1547-4204-b5cd-f12238f75032" />

### Does it work in dual JVM mode?

```
$ ENSO_LAUNCHER=native sbt --java-home ~/bin/graalvm-community-openjdk-25+25.1/ clean buildEngineDistribution
```
- After this command, there is no `.so` for `ydoc-server` 
```
$ find built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/component/ | grep so$
built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/component/libjlinenative.so
built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/component/libenso_parser.so
```
- one has to build it explicitly:
```
sbt:enso> ydoc-server/buildNativeImage
```
then the `.so` file for #14232 `ydoc-server` appears:
```
$ find built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/component/ | grep so$
built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/component/libjlinenative.so
built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/component/libenso_parser.so
built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/component/org.enso.ydoc.server.so
```
- fixed in 88c0bd0807 and also #14254

### Conversions between `Table` from different JVMs

- with clean rebuild
- one can then execute `JAVA_HOME=~/bin/graalvm-community-openjdk-25+25.1/ corepack pnpm run dev:gui`
- and get following errors

<img width="1015" height="482" alt="obrazek" src="https://github.com/user-attachments/assets/ad2cbb70-cba8-486a-8b69-7a79923e2c15" />

- the location/stacktrace of `Cannot convert 'org.enso.table.data.table.Table@8d8f754'(language: Enso Polyglot Bridge, type: org.enso.table.data.table.Table) to Java type 'org.enso.table.data.table.Table': Unsupported target type.`  isn't printed anywhere...

JaroslavTulach committed:
Build ydoc-server shared library when shouldBuildNativeImage

JaroslavTulach commented:
> ### Conversions between `Table` from different JVMs
> 
> <img alt="obrazek" width="1015" height="482" src="https://private-user-images.githubusercontent.com/26887752/510674985-ad2cbb70-cba8-486a-8b69-7a79923e2c15.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjI0MjMxMTYsIm5iZiI6MTc2MjQyMjgxNiwicGF0aCI6Ii8yNjg4Nzc1Mi81MTA2NzQ5ODUtYWQyY2JiNzAtY2JhOC00ODZhLThiNjktN2E3OTkyM2UyYzE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTA2VDA5NTMzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIyYjE5ZDFjMDZiYTU5OGU0YTg0MzY1NzU0MjM2MjJlNmJmMDczZmZmZDVjYTg3Njc5ZTljMGNlZGFmZDczMjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MvjmkSH5u76q5LQVKxjjCb8SUnup_rQIvgjei_LtwQ0">
> 
>     * the location/stacktrace of `Cannot convert 'org.enso.table.data.table.Table@8d8f754'(language: Enso Polyglot Bridge, type: org.enso.table.data.table.Table) to Java type 'org.enso.table.data.table.Table': Unsupported target type.`  isn't printed anywhere...

The same problem can be reproduced with "shell" launchers, but running in "mock dual JVM" mode:
```
JAVA_TOOL_OPTIONS=-Dpolyglot.enso.classLoading=Standard.Google:guest,hosted corepack pnpm run dev:gui
```
this can be debugged and shows that:
- error from [DataQualityMetrics.triggerTable java_table](https://github.com/enso-org/enso/blob/aa8a3234032575ad7fe8fd788965001b4604e7a3/distribution/lib/Standard/Table/0.0.0-dev/src/In_Memory_Table.enso#L91)
- fixed in 0285457

JaroslavTulach committed:
Allow passing in JVM options like -Dpolyglot.enso.classLoading=

JaroslavTulach committed:
Avoid computing DataQualityMetrics for Standard.Google tables

JaroslavTulach committed:
Merge remote-tracking branch 'origin/develop' into wip/jtulach/DualGoogle

JaroslavTulach commented:
- There is a failure with most recent `develop` merged into this PR
- The failure happens as soon as `guest-code-2` tries to send first message
```
...
app/gui dev:vite: thread: guest-code-1 sends message: OtherJvmMessage[id=19, message=Message[boolean com.oracle.truffle.api.interop.InteropLibrary.isNumber(Object)(Object), args=[]]
app/gui dev:vite: thread: guest-code-1 sends message: OtherJvmMessage[id=19, message=Message[boolean com.oracle.truffle.api.interop.InteropLibrary.hasMetaObject(Object)(Object), args=[]]
app/gui dev:vite: thread: guest-code-1 sends message: OtherJvmMessage[id=19, message=Message[Object com.oracle.truffle.api.interop.InteropLibrary.getMetaObject(Object)(Object), args=[]]
app/gui dev:vite: thread: guest-code-1 sends message: OtherJvmMessage[id=19, message=Message[boolean com.oracle.truffle.api.interop.InteropLibrary.hasMetaObject(Object)(Object), args=[]]
app/gui dev:vite: thread: guest-code-1 sends message: OtherJvmMessage[id=19, message=Message[Object com.oracle.truffle.api.interop.InteropLibrary.getMetaObject(Object)(Object), args=[]]
app/gui dev:vite: thread: guest-code-1 sends message: OtherJvmMessage[id=13, message=Message[Object com.oracle.truffle.api.interop.InteropLibrary.getMetaSimpleName(Object)(Object), args=[]]
app/gui dev:vite: thread: guest-code-2 sends message: OtherJvmMessage[id=7, message=Message[boolean com.oracle.truffle.api.interop.InteropLibrary.isMemberInvocable(Object,String)(Object, String), args=[DEFAULT_COLUMN_NAMES]]
````
- this is probably caused by [allowCoreThreadsTimeout](https://github.com/enso-org/enso/pull/13907/files#r2503655230)
- e.g. the `Channel` isn't yet ready for multiple threads access

JaroslavTulach committed:
Detect multi threaded access to Channel and/or JVM

JaroslavTulach committed:
Merging with develop's support for multiple threads

JaroslavTulach committed:
Remember exception during polyglotJava initialization

JaroslavTulach commented:
- Support for `GoogleSheets` credentials has been restored
- @AdRiley sample project seems to work for me both in `enso-linux-x86_64-2025.3.1-nightly.2025.11.9.AppImage` and the version in this PR

<img width="2183" height="733" alt="obrazek" src="https://github.com/user-attachments/assets/704d9598-5940-4ce2-9453-cd337947f228" />

- testing [night build from this branch](https://github.com/enso-org/enso/actions/runs/19256839216)...
- ...running `./enso-linux-x86_64-2025.3.1-nightly.2025.11.11.1.AppImage` also works for me!

JaroslavTulach commented on build.sbt:
```diff
@@ -4063,9 +4063,6 @@ lazy val `engine-runner` = project
           `database-polyglot-root`
             .listFiles("*.jar")
             .map(_.getAbsolutePath()) ++
-          `google-polyglot-root`
```
- The main change of this PR
- `std-bits/google` are no longer included when building `enso` NI

JaroslavTulach commented on std-bits/table/src/main/java/org/enso/table/read/ExcelReader.java:
```diff
@@ -430,7 +432,11 @@ private static Table readTable(
+      var emptySheetError = emptySheetType.invokeMember("Error");
+      var errorType = EnsoMeta.getType("Standard.Base.Error", "Error");
+      var error = errorType.invokeMember("throw", emptySheetError);
+      throw error.throwException();
```
- @jdunkerley, @AdRiley: I'd like to propose this as a new way to throw Enso error from Java code
- construct the right `payload` - e.g. `Empty_Sheet.Error`
- construct _dataflow error_ object by calling `Error.throw`
- and either return the _dataflow error_
- or `throwException()` it and `Panic.recover` [later](https://github.com/enso-org/enso/pull/14040/files#r2517305302) 
- just [simply recover](https://github.com/enso-org/enso/pull/14040/files#r2517305302) - no need for any conversions of `EmptySheetException` Java exception to Enso payload

jdunkerley replied:
I really don't like making these classes Enso dependent - we debug them a lot in pure Java.
Particularly the Excel code if complex and hard to work with.

JaroslavTulach replied:
- debugging in _"pure Java"_ while referencing Enso language is possible
- demonstrated by #14287

JaroslavTulach commented on distribution/lib/Standard/Table/0.0.0-dev/src/Errors.enso:
```diff
@@ -421,7 +420,7 @@ type Empty_Sheet
        ---
     handle_java_exception =
-        Panic.catch EmptySheetException handler=(_ -> Error.throw Empty_Sheet.Error)
+        Panic.recover Empty_Sheet
```
- converts a `Panic` with `Empty_Sheet` payload
- to _dataflow_ `Error`

jdunkerley replied:
As commented below would rather not change Excel code for Google sheets.

JaroslavTulach replied:
Reverted back in ec9abedc5f5bea1369812c5a6e6659d7e5e1327f

JaroslavTulach commented:
> ### Conversions between `Table` from different JVMs
> 
> 
> <img alt="obrazek" width="1015" height="482" src="https://private-user-images.githubusercontent.com/26887752/510674985-ad2cbb70-cba8-486a-8b69-7a79923e2c15.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjI5MzUyMzAsIm5iZiI6MTc2MjkzNDkzMCwicGF0aCI6Ii8yNjg4Nzc1Mi81MTA2NzQ5ODUtYWQyY2JiNzAtY2JhOC00ODZhLThiNjktN2E3OTkyM2UyYzE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTEyVDA4MDg1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFhZTQ3MWY5ZTJlNmE4NWZmNWMyZDZkN2I0MjgzNWQ0M2U2MjY5ZGY2NDBjZGI2ODQzYWFlM2I0NjQ3Yjg4YjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.97Y4uGco_5iF2xdam3qq9c8ueXDZAscXDDH5yaNEbz4">
> 
>     * the location/stacktrace of `Cannot convert 'org.enso.table.data.table.Table@8d8f754'(language: Enso Polyglot Bridge, type: org.enso.table.data.table.Table) to Java type 'org.enso.table.data.table.Table': Unsupported target type.`  isn't printed anywhere...

- these conversion problems seem to be gone
- I believe @jdunkerley's #14263 plays a significant role here
- by associating a `Table` with its JVM and selecting the right classes to invoke from the right JVM
- the convertions are no longer happening and the correct JVM is consulted to provide the result

jdunkerley commented on distribution/lib/Standard/Google/0.0.0-dev/src/Google_Sheets_Workbook.enso:
```diff
@@ -167,14 +169,14 @@ type Google_API_Error
         self.details.at 'message'
 
-private _handle_exceptions ~action = Empty_Sheet.handle_java_exception <|
+private _handle_exceptions ~action =
```
We could inline the Empty_Sheet handler here and then the exception will be from correct JVM.

JaroslavTulach replied:
The handler is inlined in 1ebffbb05ad

JaroslavTulach commented on std-bits/table/src/main/java/org/enso/table/error/EmptySheetException.java:
```diff
@@ -1,7 +0,0 @@
-package org.enso.table.error;
-
-public class EmptySheetException extends RuntimeException {
```
A pair Java class to Enso's `Standard.Table.Errors.Empty_Sheet`

JaroslavTulach commented:
Re. [this comment](https://github.com/enso-org/enso/pull/14040#discussion_r2517767465):

> I really don't like making these classes Enso dependent - we debug them a lot in pure Java.
> Particularly the Excel code if complex and hard to work with.

- debugging in _"pure Java"_ is possible even when referencing Enso 
   - #14287 demonstrates the idea
   - if there are issues with `ContextUtils` approach, we can address them
- however the cost of duplicating every error and exception cannot really be addressed
   - [EmptySheetException](https://github.com/enso-org/enso/pull/14040/files#r2518318783) & `Empty_Sheet` isn't the only example
   - there is a ton of such _"exception/error pairs"_ all over the codebase
- I don't think it is a good coding pattern
   - it is very verbose - tons of lines in Java and many in Enso
   - it spreads `handle_xyz_exception` methods all over unrelated places
   - chance of forgetting to catch the Java exception somewhere isn't zero
   - simpler to just create the Enso `Error` and/or `Panic` directly
- moreover `Error.throw` or `Panic.throw` will play much more nicely with dual JVM mode

JaroslavTulach committed:
Reveting back changes related to Error.throw being used from Java

JaroslavTulach committed:
Convert google's EmptySheetException to Empty_Sheet.Error

JaroslavTulach committed:
Merge remote-tracking branch 'origin/develop' into wip/jtulach/DualGoogle

jdunkerley reviewed (approved):
Approving the libs changes.

jdunkerley commented:
>   * [Testing EnsoMeta ability to load Enso types from Standard.Base #14287](https://github.com/enso-org/enso/pull/14287) demonstrates the idea

I think with your solution we can move in this direction.

We can still keep the JUnit tests within the std-bits libraries and keep the tight code cycle I've been there when needed. 
I agree having error/exception is not nice code pattern and we end up having it everywhere as this is the only solution we had.

I'll try out the approach and see how it goes going forward.

JaroslavTulach committed:
Pass in credential_reference as a factory class

JaroslavTulach committed:
Clear also polyglotBindings when closing a channel

JaroslavTulach committed:
Merging with most recent develop changes

4e6 commented on lib/java/jvm-interop/src/main/java/org/enso/jvm/interop/impl/OtherJvmMessage.java:
```diff
@@ -191,4 +191,14 @@ public Void apply(Channel<OtherJvmPool> t) {
   }
+
+  /** Requests polyglot bindings from the other side. */
+  @Persistable(id = 81912)
```
Shouldn't the serial version of persistence map  be also updated? https://github.com/enso-org/enso/blob/f66930ef483a1b0136472f7a36d9c558fa46cd24/lib/java/persistance/src/main/java/org/enso/persist/PerMap.java#L9

JaroslavTulach replied:
No, that's not necessary. The system will add the `id=81912` automatically to the checksum. Moreover the Channel Persistance isn't persisted to disk. Just exchanged inside of a single process. As such no versioning problem can happen anyway.
